modules/library/render/src/test/java/org/geotools/renderer/lite/StreamingRendererTest.java

```diff
@@ -927,4 +933,136 @@ public void errorOccurred(Exception e) {
         mapContent.dispose();
         assertEquals(0, errors);
     }
+
+    /**
+     * Test that StreamingRenderer#findStyleAttributes correctly extracts the base property for
+     * xpath expressions
+     */
+    @Test
+    public void testFindStyleAttributeWithXPath() throws Exception {
+        final List<String[]> properties = new ArrayList<>();
+
+        SimpleFeatureTypeBuilder builder = new SimpleFeatureTypeBuilder();
+        builder.setName("xpath");
+        builder.add("geom", LineString.class, WGS84);
+        builder.userData("xml", Boolean.TRUE);
+        builder.add("complex", String.class);
+        SimpleFeatureType complexType = builder.buildFeatureType();
+
+        DefaultFeatureCollection fc = new DefaultFeatureCollection();
+        SimpleFeature f0 =
+                SimpleFeatureBuilder.build(
+                        complexType,
+                        new Object[] {gf.createLineString(), "<props><color>blue</color></props>"},
+                        null);
+        fc.add(f0);
+
+        SimpleFeatureSource testSource =
+                new CollectionFeatureSource(fc) {
+                    @Override
+                    public SimpleFeatureCollection getFeatures(Query query) {
+                        properties.add(query.getPropertyNames());
+                        return super.getFeatures(query);
+                    }
+                };
+
+        StyleBuilder sb = new StyleBuilder();
+        // add a complex expression to the style
+        Expression color = StreamingRenderer.filterFactory.property("/complex/props/color");
```

Comment:
So the actual attribute structure of the feature type has a "complex" property that is a string, which is marked to be "xml" using a user attribute in the descriptor, and then there  is a property extractor that can drill into the string and extract bits out of it. That is a "creative" use of property accessors, no wonder StreamingRenderer was not prepared for it. We can access JSON structures stored as strings in a similar way using a filter function, that required no changes in the rest of the code.

Reply:
Yes, you captured it perfectly. I believe we came up with this approach based on some of the existing geotools code, but it's been a while and I no longer recall exactly what.
Are there any examples of the existing json filter function you could link to? We also have a jsonPath filter function, but at least in our implementation, we use the same property name expression so I think we hit the same issue (i.e. `jsonPath('$.foo.bar')`. Maybe we could rewrite the filter function to take `jsonPath('foo', 'bar')` instead.

Reply:
Yes, the filter function should take attributes that are visible in the feature type, something that can be described as is, that can be given a descriptor of its own (see my question about the property accessor).

The function I was referring to is a JSONPointer one, a quick GitHub search provides references and examples:
https://github.com/geotools/geotools/search?q=jsonpointer

Reply:
thanks, based on that function, it seems like separating out the path from the property name would probably work for us if this PR is ultimately deemed undesirable.